pythonFiles/include/blender_vscode/load_addons.py

```diff
@@ -1,67 +1,216 @@
 import os
+import subprocess
 import sys
 import traceback
 from pathlib import Path
-from typing import List
+from typing import List, Union, Optional, Dict
 
 import bpy
 
 from . import AddonInfo
 from .communication import send_dict_as_json
 from .environment import addon_directories
-from .utils import is_addon_legacy
+from .utils import is_addon_legacy, addon_has_bl_info
 
+if bpy.app.version >= (4, 2, 0):
+    _EXTENSIONS_DEFAULT_DIR = Path(bpy.utils.user_resource("EXTENSIONS", path="user_default"))
+else:
+    _EXTENSIONS_DEFAULT_DIR = None
+_ADDONS_DEFAULT_DIR = Path(bpy.utils.user_resource("SCRIPTS", path="addons"))
 
-def setup_addon_links(addons_to_load: List[AddonInfo]):
 
-    path_mappings = []
+def setup_addon_links(addons_to_load: List[AddonInfo]) -> List[Dict]:
+    path_mappings: List[Dict] = []
 
-    for addon_info in addons_to_load:
-        user_addon_directory = get_user_addon_directory(Path(addon_info.load_dir))
-        print(f"USER ADDON: {user_addon_directory}")
-
-        if not os.path.exists(user_addon_directory):
-            os.makedirs(user_addon_directory)
+    # always make sure addons are in path, important when running fresh blender install
+    # do it always to avoid very confusing logic in the loop below
+    os.makedirs(_ADDONS_DEFAULT_DIR, exist_ok=True)
+    if str(_ADDONS_DEFAULT_DIR) not in sys.path:
+        sys.path.append(str(_ADDONS_DEFAULT_DIR))
 
-        if is_addon_legacy(Path(addon_info.load_dir)) and not str(user_addon_directory) in sys.path:
-            sys.path.append(str(user_addon_directory))
+    remove_broken_addon_links()
+    if bpy.app.version >= (4, 2, 0):
+        remove_broken_extension_links()
 
-        if is_in_any_addon_directory(addon_info.load_dir):
-            load_path = addon_info.load_dir
+    for addon_info in addons_to_load:
+        try:
+            load_path = _link_addon_or_extension(addon_info)
+        except PermissionError as e:
+            print(
+                f"""ERROR: {e} 
+Path "{e.filename}" can not be removed. **Please remove it manually!** Most likely causes:
+    - Path requires admin permissions to remove
+    - Windows only: You upgraded Blender version and imported old setting. Now links became real directories.
+    - Path is a real directory with the same name as addon (removing might cause data loss!)"""
+            )
+            raise e
```

Comment:
technically, there is no need to "panic: here, we can still load other addons, but there is no indication that linking failed and bledner needs to be restarted anyways...